win32/tests/call_once_int/call_once_int.c

```diff
@@ -0,0 +1,165 @@
+
+static DWORD WINAPI sleepThread(
+    _In_ LPVOID lpParameter
+)
+{
+    (void)lpParameter;
+    while (call_once_begin(&g_stateSleep) == CALL_ONCE_PROCEED)
+    {
+        (void)interlocked_increment(&threadThatSleeps_executions);
+        Sleep(4000);
+        call_once_end(&g_stateSleep, true);
+    }
+    return 0;
+}
+
+#define N_THREADS_FOR_CHAOS 16
+#define N_AT_LEAST_TIME_MS 2000 /*ms*/
+static volatile_atomic int32_t n_threads_that_failed = 0;
+static double startTime;
+static volatile_atomic int32_t chaosThread_executions = 0;
+static volatile_atomic int32_t g_stateChaos = CALL_ONCE_NOT_CALLED;
+
+static DWORD WINAPI chaosThread(
+    _In_ LPVOID lpParameter
+)
+{
+    (void)lpParameter;
+    bool shouldIncrementThreadsThatFailed = true;
+    while (call_once_begin(&g_stateChaos) != CALL_ONCE_ALREADY_CALLED)
+    {
+        if (
+            (timer_global_get_elapsed_ms() - startTime <= N_AT_LEAST_TIME_MS) ||
+            (interlocked_add(&n_threads_that_failed, 0) != N_THREADS_FOR_CHAOS)
+            )
+        {
+            /*continue to fail*/
+            call_once_end(&g_stateChaos, false);
+
+            if (shouldIncrementThreadsThatFailed)
+            {
+                (void)interlocked_increment(&n_threads_that_failed);
+                shouldIncrementThreadsThatFailed = false;
+            }
+        }
+        else
+        {
+            /*now that enough time has passed... and at least every thread has failed at least once... end of the threads*/
+            call_once_end(&g_stateChaos, true);
+            (void)interlocked_increment(&chaosThread_executions);
+        }
+    }
+    return 0;
+}
+
+BEGIN_TEST_SUITE(call_once_inttests)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+    test_serialize_mutex = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(test_serialize_mutex);
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+    TEST_MUTEX_DESTROY(test_serialize_mutex);
+}
+
+TEST_FUNCTION_INITIALIZE(method_init)
+{
+    if (TEST_MUTEX_ACQUIRE(test_serialize_mutex))
+    {
+        ASSERT_FAIL("Could not acquire test serialization mutex.");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(method_cleanup)
+{
+    TEST_MUTEX_RELEASE(test_serialize_mutex);
+}
+
+/*Tests_SRS_CALL_ONCE_02_003: [ If interlocked_compare_exchange returns 1 then call_once_begin shall call InterlockedHL_WaitForNotValue(state, 1, INFINITE) and call again interlocked_compare_exchange(state, 1, 0). ]*/
+TEST_FUNCTION(call_once_will_wake_a_waiting_thread)
+{
+    /*this test spawns 2 threads*/
+    /*both threads are trying to compete over the same call_once_sequence*/
+    /*the sequence has a sleep(4000); */
+    /*it is assumed at this time that the first thread will get to its sleep*/
+    /*the second thread will be stucked in a "waitfornotvalue"*/
+
+    /*the first thread will exit succesfully, thus unblocking the second thread*/
+
+    ///arrange
+    HANDLE threads[2];
+
+    ///act
+    threads[0] = CreateThread(NULL, 0, sleepThread, NULL, 0, NULL);

```

Comment:
>CreateThread [](start = 17, length = 12)

thread_api for future PAL friendliness?

Reply:
I wish I could do that, but thread_api doesn't really have the support to join multiple threads at the same time, which is how it is used here [WaitForMultipleObjects] (and below). Changing it would bring better future compatibility, but I would believe that future compatibility is really a future endeavor. Who knows, if I dwell enough in azure-c-pal I might end up doing it myself!


---
In reply to: [466043191](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466043191) [](ancestors = 466043191)